util/perl/TLSProxy/Proxy.pm

```diff
@@ -193,31 +194,49 @@ sub start
         return 0;
     }
 
-    $pid = fork();
-    if ($pid == 0) {
-        my $execcmd = $self->execute
-            ." s_server -max_protocol TLSv1.3 -no_comp -rev -engine ossltest -accept "
-            .($self->server_port)
-            ." -cert ".$self->cert." -cert2 ".$self->cert
-            ." -naccept ".$self->serverconnects;
-        unless ($self->supports_IPv6) {
-            $execcmd .= " -4";
-        }
-        if ($self->ciphers ne "") {
-            $execcmd .= " -cipher ".$self->ciphers;
-        }
-        if ($self->ciphersuitess ne "") {
-            $execcmd .= " -ciphersuites ".$self->ciphersuitess;
-        }
-        if ($self->serverflags ne "") {
-            $execcmd .= " ".$self->serverflags;
-        }
-        if ($self->debug) {
-            print STDERR "Server command: $execcmd\n";
-        }
-        exec($execcmd);
+    my $execcmd = $self->execute
+        ." s_server -max_protocol TLSv1.3 -no_comp -rev -engine ossltest"
+        ." -accept 0 -cert ".$self->cert." -cert2 ".$self->cert
+        ." -naccept ".$self->serverconnects;
+    unless ($self->supports_IPv6) {
+        $execcmd .= " -4";
+    }
+    if ($self->ciphers ne "") {
+        $execcmd .= " -cipher ".$self->ciphers;
+    }
+    if ($self->ciphersuitess ne "") {
+        $execcmd .= " -ciphersuites ".$self->ciphersuitess;
+    }
+    if ($self->serverflags ne "") {
+        $execcmd .= " ".$self->serverflags;
+    }
+    if ($self->debug) {
+        print STDERR "Server command: $execcmd\n";
+    }
+
+    $self->serverpid(open(my $sh, "$execcmd |"));
+
+    # Process the output from s_server until we find the ACCEPT line, which
+    # tells us what the accepting address and port are.
+    while (<$sh>) {
+        print;
+        s/\R$//;                # Better chomp
+        next unless (/^ACCEPT\s.*:(\d+)$/);
+        $self->{server_port} = $1;
+        last;
     }
-    $self->serverpid($pid);
+
+    # Just make sure everything else is simply printed.
+    # The sub process simply inherits our STDIN and will keep consuming
+    # and printing it as long as there is anything there, out of our way.
+    if (fork() == 0) {
+        open STDIN, '<&', $sh;
+        close $sh;
+        exec ("$^X -ne print");
+    }
```

Comment:
Thing about fork() is that there is no one on Windows. fork you see in MSWin32 perl creates a thread. This means that whatever you do in {} affects global variables in current script. This can have fatal consequences. Another problem is that you probably want to synchronize on termination of created process, while what you get from Windows "fork" is [masqueraded] thread id. So to summarize. Take *everything* out of if (fork() == 0) {} and don't use fork on Windows. But what are the options on Windows then? Win32::Process::Create. It will return synchronizable object that one can use in place for wait[pid]. In other words there should be if (eval { require Win32::Process; 1; }) and switch between fork&exec and Win32::Process::Create.

Just in case, you can wonder but how does it work in MSYS2 or Cygwin? Those are running on Windows, but have no Win32::Process. Well, I'm not sure about MSYS2, but what Cygwin does is create *suspended* process with current image, copy own working set to the created process, still suspended, do some other preparations, modify magic location in the copy that makes it jump to exit point from fork as the process gets unsuspended. I would guess MSYS2 does the same...

Reply:
> Win32::Process::Create. It will return synchronizable object that one can use in place for wait[pid]

Though it should be possible to pull process id from the object and still use waitpid...